File: backend/app/schemas/ledger.py
from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel

from app.models.ledger_entry import Currency, EntryDirection
from app.models.transaction import TransactionType


# ── Posición financiera ──────────────────────────────────────────────────────

class BalancePosition(str, Enum):
    """
    Interpretación semántica del saldo neto de un cliente en una divisa.

    Convención de signo (perspectiva empresa — accounts receivable):
      raw_balance > 0  →  CLIENT_OWES   el cliente nos debe esa cantidad
      raw_balance < 0  →  COMPANY_OWES  le debemos al cliente esa cantidad
      raw_balance == 0 →  SETTLED       cuenta saldada
    """
    CLIENT_OWES = "CLIENT_OWES"    # el cliente me debe
    COMPANY_OWES = "COMPANY_OWES"  # yo le debo al cliente
    SETTLED = "SETTLED"            # saldo cero — saldado


class CurrencyBalance(BaseModel):
    """Posición financiera de un cliente en una sola divisa."""
    raw_balance: Decimal
    """Saldo con signo: positivo = debe al negocio, negativo = negocio le debe."""
    absolute_balance: Decimal
    """Valor absoluto del saldo, siempre >= 0."""
    position: BalancePosition
    """Interpretación financiera del signo."""
    display_label: str
    """Etiqueta legible para UI: 'El cliente debe' | 'A favor del cliente' | 'Saldado'."""


# ── Saldo agregado del cliente ───────────────────────────────────────────────

class BalanceOut(BaseModel):
    client_id: UUID
    mxn: CurrencyBalance
    gtq: CurrencyBalance

    # ── Equivalencias referenciales ─────────────────────────────────────────
    # IMPORTANTE: estos campos son puramente visuales. NO son saldo oficial.
    # Solo se calculan si el consumidor envía reference_exchange_rate.
    # La tasa la provee el operador; el sistema no tiene tasa automática.
    equivalent_in_mxn: Optional[Decimal] = None
    """Saldo GTQ convertido a MXN a la tasa de referencia. Solo referencial."""
    equivalent_in_gtq: Optional[Decimal] = None
    """Saldo MXN convertido a GTQ a la tasa de referencia. Solo referencial."""
    reference_exchange_rate: Optional[Decimal] = None
    """
    Tasa enviada por el consumidor (MXN por 1 GTQ).
    Convención: reference_exchange_rate = amount_mxn / amount_gtq.
    Solo para equivalencia visual — no afecta el saldo oficial.
    """


# ── Libro mayor ──────────────────────────────────────────────────────────────

class LedgerEntryOut(BaseModel):
    """
    Una fila del libro mayor enriquecida con contexto de la transacción.

    Semántica de direction para la UI:
      CREDIT (+) → dinero que ENTRA al cliente en esa divisa
      DEBIT  (−) → dinero que SALE del cliente en esa divisa

    Solo aparecen entradas de transacciones ACTIVE.
    Las transacciones anuladas se excluyen porque no afectan el saldo real.
    """
    id: UUID
    transaction_id: UUID
    transaction_code: str           # e.g. "TXN-20260401-00001"
    transaction_type: TransactionType  # e.g. SELL_MXN, PAYMENT, etc.
    currency: Currency              # MXN | GTQ
    direction: EntryDirection       # CREDIT (+) | DEBIT (−)
    amount: Decimal
    balance_after: Decimal          # saldo snapshot tras este movimiento
    created_at: datetime
    notes: Optional[str] = None     # notas de la transacción origen, si existen


class LedgerResponse(BaseModel):
    client_id: UUID
    total: int                      # total de entradas en el filtro (para paginación)
    skip: int
    limit: int
    entries: List[LedgerEntryOut]
    balance: BalanceOut             # saldo actual (siempre calculado, nunca almacenado)


# ── Ledger agrupado por transacción ──────────────────────────────────────────
#
# Diferencia con LedgerResponse (detallado):
#   - LedgerResponse: una fila por ledger_entry (una por divisa por movimiento).
#     Dos movimientos de distinta divisa de la misma transacción → dos filas.
#   - LedgerGroupedResponse: una fila por transacción, con columnas separadas
#     para egreso/ingreso por divisa. Ideal para vista tipo Excel.
#
# Convención egreso / ingreso (perspectiva empresa):
#   egreso  = CREDIT en ledger (empresa dio al cliente)
#   ingreso = DEBIT  en ledger (empresa recibió del cliente)
#
# Convención net (perspectiva cuentas por cobrar):
#   net = ingreso - egreso
#   net > 0  → CLIENT_OWES  (empresa recibió más de lo que dio → cliente debe)
#   net < 0  → COMPANY_OWES (empresa dio más de lo que recibió → le debemos al cliente)
#   net = 0  → SETTLED

class LedgerGroupedRow(BaseModel):
    """Una fila por transacción en la vista agrupada del libro mayor."""
    transaction_id: UUID
    date: datetime
    reference: str                  # código de transacción, e.g. TXN-20260421-00001
    operation: TransactionType
    egreso_mxn: Decimal             # MXN que la empresa dio al cliente (CREDIT MXN)
    egreso_gtq: Decimal             # GTQ que la empresa dio al cliente (CREDIT GTQ)
    ingreso_mxn: Decimal            # MXN que la empresa recibió del cliente (DEBIT MXN)
    ingreso_gtq: Decimal            # GTQ que la empresa recibió del cliente (DEBIT GTQ)
    notes: Optional[str] = None
    status: str


class LedgerGroupedSummary(BaseModel):
    """
    Totales y posición neta para el rango consultado (no afectado por paginación).

    net_mxn = total_ingreso_mxn - total_egreso_mxn
    net_gtq = total_ingreso_gtq - total_egreso_gtq

    Etiquetas:
      CLIENT_OWES  → "El cliente debe"      (net > 0)
      COMPANY_OWES → "A favor del cliente"  (net < 0)
      SETTLED      → "Saldado"              (net = 0)
    """
    total_egreso_mxn: Decimal
    total_ingreso_mxn: Decimal
    total_egreso_gtq: Decimal
    total_ingreso_gtq: Decimal
    net_mxn: Decimal
    net_gtq: Decimal
    net_mxn_label: str
    net_gtq_label: str


class LedgerGroupedResponse(BaseModel):
    """
    Vista agrupada del libro mayor: una fila por transacción.
    El summary refleja el rango completo de fechas; las rows están paginadas.
    """
    client_id: UUID
    total: int
    skip: int
    limit: int
    summary: LedgerGroupedSummary
    rows: List[LedgerGroupedRow]

File: frontend/src/types/ledger.ts
/**
 * Tipos del contrato de Libro Mayor y Balance — Kambio
 * Espejo del backend: app/schemas/ledger.py
 */

// ── Posición financiera ──────────────────────────────────────────────────────

/** Interpretación semántica del saldo neto de un cliente en una divisa. */
export type BalancePosition = 'CLIENT_OWES' | 'COMPANY_OWES' | 'SETTLED'

/** Posición financiera de un cliente en una sola divisa. */
export interface CurrencyBalance {
  /** Saldo con signo: positivo = cliente nos debe, negativo = nosotros le debemos. */
  raw_balance: number
  /** Valor absoluto del saldo, siempre >= 0. */
  absolute_balance: number
  /** Interpretación semántica del signo. */
  position: BalancePosition
  /** Etiqueta legible para UI: "El cliente debe" | "A favor del cliente" | "Saldado" */
  display_label: string
}

/** Saldo agregado del cliente por divisa. */
export interface BalanceOut {
  client_id: string
  mxn: CurrencyBalance
  gtq: CurrencyBalance
  /** Saldo GTQ convertido a MXN a la tasa referencial. Solo visual. */
  equivalent_in_mxn?: number | null
  /** Saldo MXN convertido a GTQ a la tasa referencial. Solo visual. */
  equivalent_in_gtq?: number | null
  /** Tasa enviada por el operador (MXN por 1 GTQ). No es saldo oficial. */
  reference_exchange_rate?: number | null
}

// ── Libro mayor ──────────────────────────────────────────────────────────────

export interface LedgerEntryOut {
  id: string
  transaction_id: string
  transaction_code: string
  transaction_type: string
  currency: 'MXN' | 'GTQ'
  direction: 'CREDIT' | 'DEBIT'
  amount: number
  balance_after: number
  created_at: string
  notes?: string | null
}

export interface LedgerResponse {
  client_id: string
  total: number
  skip: number
  limit: number
  entries: LedgerEntryOut[]
  balance: BalanceOut
}

// ── Ledger agrupado (nuevo endpoint: /ledger-summary) ─────────────────────────
// Una fila por transacción. Perspectiva empresa:
//   egreso  = CREDIT en ledger (empresa dio al cliente)
//   ingreso = DEBIT  en ledger (empresa recibió del cliente)

export interface LedgerGroupedRow {
  transaction_id: string
  date: string
  reference: string
  operation: string
  egreso_mxn: string | number
  egreso_gtq: string | number
  ingreso_mxn: string | number
  ingreso_gtq: string | number
  notes?: string | null
  status: string
}

export interface LedgerGroupedSummary {
  total_egreso_mxn: string | number
  total_ingreso_mxn: string | number
  total_egreso_gtq: string | number
  total_ingreso_gtq: string | number
  /** net = ingreso - egreso. Positivo → CLIENT_OWES, negativo → COMPANY_OWES */
  net_mxn: string | number
  net_gtq: string | number
  /** 'CLIENT_OWES' | 'COMPANY_OWES' | 'SETTLED' */
  net_mxn_label: string
  net_gtq_label: string
}

export interface LedgerGroupedResponse {
  client_id: string
  total: number
  skip: number
  limit: number
  /** Totales del rango completo (independiente de paginación) */
  summary: LedgerGroupedSummary
  /** Filas paginadas, 1 por transacción */
  rows: LedgerGroupedRow[]
}

/** Texto legible para net_mxn_label / net_gtq_label */
export function netLabelText(label: string): string {
  const m: Record<string, string> = {
    CLIENT_OWES:  'El cliente debe',
    COMPANY_OWES: 'A favor del cliente',
    SETTLED:      'Saldado',
  }
  return m[label] || label
}

/** Clase CSS según net label */
export function netLabelColor(label: string): string {
  if (label === 'CLIENT_OWES')  return 'text-emerald-600 dark:text-emerald-400'
  if (label === 'COMPANY_OWES') return 'text-red-600 dark:text-red-400'
  return 'text-muted'
}

/** Border class según net label (para tarjetas destacadas) */
export function netBorderClass(label: string): string {
  if (label === 'CLIENT_OWES')  return 'border-emerald-300 dark:border-emerald-800'
  if (label === 'COMPANY_OWES') return 'border-red-300 dark:border-red-800'
  return 'border-border'
}


// ── Helpers UI ───────────────────────────────────────────────────────────────

/** Clases CSS según la posición financiera del cliente. */
export function positionColorClass(position: BalancePosition): string {
  switch (position) {
    case 'CLIENT_OWES':  return 'text-emerald-600 dark:text-emerald-400'  // verde — el cliente nos debe
    case 'COMPANY_OWES': return 'text-red-600 dark:text-red-400'           // rojo — nosotros le debemos
    case 'SETTLED':      return 'text-muted'                               // gris — saldado
  }
}

/** Badge de estado según posición. */
export function positionBadgeClass(position: BalancePosition): string {
  switch (position) {
    case 'CLIENT_OWES':  return 'bg-emerald-50 text-emerald-700 dark:bg-emerald-950/40 dark:text-emerald-300'
    case 'COMPANY_OWES': return 'bg-red-50 text-red-700 dark:bg-red-950/40 dark:text-red-300'
    case 'SETTLED':      return 'bg-slate-100 text-slate-500 dark:bg-slate-800 dark:text-slate-400'
  }
}

/** Símbolo de divisa. */
export function currencySymbol(currency: 'MXN' | 'GTQ'): string {
  return currency === 'MXN' ? '$' : 'Q'
}

/** Formato monetario consistente. */
export function fmtAmount(val: number | string | null | undefined): string {
  if (val === null || val === undefined) return '0.00'
  return Number(val).toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 })
}
